File: c++/syntax.cpp
#include "syntax.hpp"
#include "errors.hpp"
#include <string>
#include <iostream>
using namespace std;

Object Syntax::nextObj() {
	auto obj = reader.nextObj();
	// NOTE - i was originally using c++ regexes here to check for
	// integers and lambdas. after profiling and discovering they
	// were the using most of the program runtime, i changed to parsing
	// them myself. WITH regexes, runtime was 4.5x higher and memory
	// usage was 1500x (!!) larger (see commit [556839e] for the regex version)
	if(obj.isSymbol()) {
		{
			// see if it's an integer
			bool has_digits = false;
			const char *s = obj.asSymbol();
			if(*s == '+' || *s == '-') 
				++s;

			while(isdigit(*s)) {
				has_digits = true;
				++s;
			}

			// integers are just pushed to stack
			if(!*s && has_digits) {
				//cout << "MATCHED INT:" << obj.fmtStackPrint() << endl;
				auto intobj = newInt(stoi(obj.asSymbol()));
				reader.deletePrevObj();
				reader.insertPrevObj(intobj);
				return intobj;
			}
		}

		{
			// to keep parsing simple (i.e. should be easy to do without regexes),
			// floats are written like #n.nnn
			if(obj.asSymbol()[0] == '#') {
				auto fobj = newFloat(stod(obj.asSymbol()+1));
				reader.deletePrevObj();
				reader.insertPrevObj(fobj);
				return fobj;
			}
		}

		{
			if(obj.isSymbol("{")) {
				return parse_lambda();
			}
			if(obj.isSymbol("(")) {
				return parse_comment(); // returns obj AFTER comment
			}
			if(obj.isSymbol(".\"")) {
				return parse_quote_printstring();
			}
		}
	}

	// no matching syntax, so return as-is
	return obj;
}

Object Syntax::nextObjOrFail() {
	if(peekObj().isNull()) {
		throw LangError("Unexpected end of input");
	}
	return nextObj();
}

Object Syntax::prevObjOrFail() {
	if(peekPrevObj().isNull()) {
		throw LangError("Unable to find previous word");
	}
	return prevObj();
}

Object Syntax::nextSymbolOrFail() {
	auto obj = nextObjOrFail();
	if(!obj.isSymbol()) {
		throw LangError("Expecting symbol, got: " + obj.fmtStackPrint());
	}
	return obj;
}

Object Syntax::parse_comment() {
	// skip comment and return NEXT word -- assumes ( was just read
	//
	// allow nested comments
	int nesting = 1;
	while(true) {
		auto obj = reader.nextObj();
		if(obj.isSymbol(")")) {
			if(--nesting == 0) {
				// end of comment - return NEXT object
				return nextObj();
			}
		}
		else if(obj.isSymbol("(")) {
			++nesting;
		}
		else if(obj.isNull()) {
			throw LangError("Unexpected end of input in comment");
		}
	}
}

Object Syntax::parse_lambda() {
	// turn { ... } into an anonymous objlist
	//
	// the FIRST time I see { .. }, create a Lambda, then
	// REMOVE the { ... } and replace it with the Lambda. return the
	// Lambda as the parsed object.
	//
	// each SUBSEQUENT time the same code runs, the Lambda will be
	// in the objlist, so will be pushed to the stack.
	//
	// from the perspective of the user, the same thing happened both times --
	// the lambda was pushed to the stack, ready to be called, stored, etc.
	
	// "{" was just read -- read the words until }
	
	// delete the { that was just read
	reader.deletePrevObj();

	auto objlist = new ObjList();
	int nesting = 1;
	while(true) {
		auto obj = nextObjOrFail();
		// delete the { ... } as I read it -- will replace it with a tagged wordlist
		reader.deletePrevObj();
		
		if(obj.isSymbol("{")) {
			// if I find inner lambdas, just copy them for now and later when they are run, 
			// this same process will happen for them
			++nesting;
			objlist->push_back(obj);
		}
		else if(obj.isSymbol("}")) {
			if(--nesting > 0) {
				objlist->push_back(obj);
				continue;
			}
			
			// replace { .. } in source wordlist with Lambda so it only has to be parsed once
			auto _lambda = newLambda(objlist);
			reader.insertPrevObj(_lambda);
			// next time, the interpreter will see the Lambda itself, so will not come back here
			return _lambda;
		}
		else {
			objlist->push_back(obj);
		}
	}
}

Object Syntax::parse_quote_printstring() {
	// ." some string here " -- rewrite to '"some string here" .'
	string s;
	// delete ."
	reader.deletePrevObj();
	while(true) {
		auto obj = reader.nextObj(); // *NOT* Syntax::nextObj() - don't want any processing, i.e.
									// don't want numbers in middle of string to be converted to ints
		reader.deletePrevObj(); // delete each obj as I read it, will replace at end
		if(obj.isSymbol("\"")) {
			// end of string - write new code
			reader.insertPrevObj(newString(s));
			// use only builtin functions so it works even with -noinit
			reader.insertPrevObj(newSymbol("puts"));
			// (will have blank at end, due to concatenation code below, so don't need to add one)
			// since i rewrote into multiple objects (unlike the lambda case above), 
			// backup the reader to string and return it
			for(int i=0; i<2; ++i) {
				reader.prevObj();
			}
			return reader.nextObj();
		}
		else if(obj.isNull()) {
			throw LangError("Unexpected end of input inside .\"");
		}
		else if(!obj.isSymbol()) {
			throw LangError("Got non-symbol from Reader (!!):" + obj.fmtStackPrint());
		}
		else {
			// since i called reader.nextObj(), this will always be a symbol
			s += string(obj.asSymbol()) + " ";
		}
	}
}

Object Syntax::peekObj() {
	// I need the above processing to occur on peeked objects,
	// so do it this way ....
	auto obj = nextObj();
	if(!obj.isNull()) {
		reader.prevObj(); // rewind
	}
	return obj;
}

Object Syntax::prevObj() {
	return reader.prevObj();
}

Object Syntax::peekPrevObj() {
	return reader.peekPrevObj();
}

File: csharp/syntax.cs
/*
	Syntax - takes words from Reader, recognizes any syntax forms, and returns
	objects to Interpreter.
	
	Copyright (c) 2022 Frank McIngvale, see LICENSE
*/
using System;
using System.Collections.Generic;
#nullable enable

public class Syntax {

	public Reader reader;

	public Syntax() {
		reader = new Reader();
	}

	// mirrored from Reader
	public void addText(string text) { reader.addText(text); }
	public void clearAll() { reader.clearAll(); }
	public void pushObjList(List<LangObject> objs) { reader.pushObjList(objs); }
	public void popObjList() { reader.popObjList(); }
	public bool hasPushedObjLists() { return reader.hasPushedObjLists(); }
 
	public LangObject nextObjOrFail() {
		if(peekObj() == null) {
			throw new LangError("Unexpected end of input");
		}
		return nextObj()!;
	}

	public LangSymbol nextSymbolOrFail() {
		var obj = nextObjOrFail();
		if(!(obj is LangSymbol)) {
			throw new LangError("Expecting symbol but got " + obj.fmtStackPrint());
		}
		return (obj as LangSymbol)!;
	}

	public LangObject prevObjOrFail() {
		if(peekPrevObj() == null) {
			throw new LangError("Unable to find previous word");
		}
		return prevObj()!;
	}

	// see C++ version for verbose comments; minimal comments here
	public LangObject parse_lambda() {
		// turn { ... } into an anonymous wordlist
		
		// delete the { that was just read
		reader.deletePrevObj();

		var objlist = new List<LangObject>();
		int nesting = 1;
		while(true) {
			var obj = nextObj();
			if(obj == null) {
				throw new LangError("Unexpected end of input inside { .. }");
			}
			var sym = obj as LangSymbol;
			// delete the { ... } as I read it -- will replace it with a lambda reference
			reader.deletePrevObj();
			if(sym != null && sym.match("{")) {
				// if I find inner lambdas, just copy them for now and later when they are run, 
				// this same process will happen for them
				++nesting;
				objlist.Add(obj);
			}
			else if(sym != null && sym.match("}")) {
				if(--nesting > 0) {
					objlist.Add(obj);
					continue;
				}

				var _lambda = new LangLambda(objlist);
				//Console.WriteLine("NEW LAMBDA IS:");
				//foreach(var o in objlist) {
				//	Console.WriteLine(o.fmtStackPrint());
				//}

				// replace { .. } in source objlist with lambda.
				reader.insertPrevObj(_lambda);
				// return lambda as the read object. on subsequent passes, the interpreter
				// will see the lambda directly in the objlist
				return _lambda;
			}
			else {
				objlist.Add(obj);
			}
		}
	}

	public LangObject? parse_comment() {
		int nesting = 1;
		while(true) {
			var sym = reader.nextObj() as LangSymbol;
			//if(sym != null) {
			//	Console.WriteLine("COMMENT READ:" + sym.fmtStackPrint());
			//}
			if(sym == null) {
				throw new LangError("Unexpected end of input inside comment");
			}
			else if(sym.value == ")") {
				if(--nesting == 0) {
					return nextObj();
				}
			}
			else if(sym.value == "(") {
				++nesting;
			}
		}
	}

	public LangObject? parse_quote_printstring() {
		// ." some string here " -- rewrite to '"some string here" .'
		//Console.WriteLine("PARSING .\" ...");
		string s = "";
		// delete ."
		reader.deletePrevObj();
		while(true) {
			var obj = reader.nextObj(); // *NOT* Syntax::nextObj() - don't want any processing, i.e.
										// don't want numbers in middle of string to be converted to ints
			if(obj == null) {
				throw new LangError("Unexpected end of input inside .\"");
			}
			var sym = obj as LangSymbol;
			reader.deletePrevObj(); // delete each obj as I read it, will replace at end
			if(sym == null) {
				throw new LangError("Got non-symbol from Reader (!!):" + obj.fmtStackPrint());
			}
			else if(sym.match("\"")) {
				// end of string - write new code
				//Console.WriteLine("FINISHED STRING: " + s);
				reader.insertPrevObj(new LangString(s));
				// use only builtin functions so it works even with -noinit
				reader.insertPrevObj(new LangSymbol("puts"));
				// (will have blank at end, due to concatenation code below, so don't need to add one)
				// since i rewrote into multiple objects (unlike the lambda case above), 
				// backup the reader to string and return it
				for(int i=0; i<2; ++i) {
					reader.prevObj();
				}
				//reader.debug_print_objlist();
				return reader.nextObj();
			}
			else {
				// since i called reader.nextObj(), this will always be a symbol
				s += sym.value + " ";
			}
		}
	}

	// DON'T use regex here -- this is run for potentially every word so has to be fast
	public bool isInteger(string text) {
		int i=0;
		bool hasdigits=false;
		if(text[i] == '+' || text[i] == '-') {
			++i;
		}
		while(i < text.Length && text[i] >= '0' && text[i] <= '9') {
			++i;
			hasdigits=true;
		}
		return (i == text.Length && hasdigits);
	}

	public LangObject? nextObj() {
		var obj = reader.nextObj();
		if (obj is LangSymbol) {
			var sym = obj as LangSymbol;	

			// floats: #NNN.NN
			if(sym!.match("#",1)) {
				var fobj = new LangFloat(double.Parse(sym!.value.Substring(1)));
				reader.deletePrevObj();
				reader.insertPrevObj(fobj);
				return fobj;
			}	
			else if(sym!.match("{")) {
				return parse_lambda();
			}
			else if(sym!.match("(")) {
				return parse_comment();
			}
			else if(sym!.match(".\"")) {
				return parse_quote_printstring();
			}
			// integers (last since test is more expensive than those above)
			// [in practice this doesn't help too much since MOST words in a real
			// program won't match any of these rules, so all have to run for nearly every word]
			// may help more later if/when most words are NOT left as symbols but converted
			// to some other object
			else if(isInteger(sym!.value)) {
				var intobj = new LangInt(int.Parse(sym!.value));
				reader.deletePrevObj();
				reader.insertPrevObj(intobj);
				return intobj;
			}
		}
		return obj;
	}

	public LangObject? peekObj() {
		// I need the above processing to occur on peeked objects,
		// so do it this way ....
		var obj = nextObj();
		if(obj != null) {
			reader.prevObj(); // rewind
		}
		return obj;
	}

	public LangObject? prevObj() { return reader.prevObj(); }
	public LangObject? peekPrevObj() { return reader.peekPrevObj(); }

	public void debug_print_objlist() { reader.debug_print_objlist(); }
};

